src/Castle.Core/DynamicProxy/Internal/InvocationHelper.cs

```diff
@@ -18,16 +18,17 @@ namespace Castle.DynamicProxy.Internal
 	using System.Collections.Generic;
 	using System.Diagnostics;
 	using System.Reflection;
+	using System.Threading;
 
-	using Castle.Core.Internal;
 	using Castle.DynamicProxy.Generators;
 
 	public static class InvocationHelper
 	{
 		private static readonly Dictionary<CacheKey, MethodInfo> cache =
 			new Dictionary<CacheKey, MethodInfo>();
 
-		private static readonly Lock @lock = Lock.Create();
+		private static readonly ReaderWriterLockSlim cacheLock =
+			new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
```

Comment:
You often see a `Dictionary<,>` together with a `ReaderWriterLockSlim`. These could be wrapped up as an internal helper type similar to the BCL's `ConcurrentDictionary<>`, except we might only need a `GetOrAdd` method. Having such a type would mean we can replace all of the double-checked lock pattern that's currently being used in several places with a single `GetOrAdd` call. (This should work in `ProxyUtil`, `InvocationHelper`, `DictionaryAdapterFactory`; but probably not in `ModuleScope` nor `SingletonDispenser` due to the added synchronization requirements these two types have).